File: clicky-demo/ClickySDK/Sources/ClickySDK/Internal/ClaudeAPI.swift
import Foundation

struct ClaudeMessage {
    enum Role: Equatable { case user, assistant }
    let role: Role
    let text: String
}

// Minimal Claude streaming client used by the SDK.
//
// PRODUCTION NOTE: this calls api.anthropic.com directly with the key
// passed in via ClickyConfig. For real apps, point it at a proxy URL
// you control (Cloudflare Worker, your backend) so the key isn't
// embedded in the binary. The SDK boundary is designed so swapping
// the endpoint is a one-line change here.
final class ClaudeAPI {
    private let apiKey: String
    private let model: String
    private let endpoint = URL(string: "https://api.anthropic.com/v1/messages")!

    init(apiKey: String, model: String) {
        self.apiKey = apiKey
        self.model = model
    }

    enum StreamEvent {
        case textDelta(String)
        case done
        case error(Error)
    }

    func stream(
        systemPrompt: String,
        conversation: [ClaudeMessage],
        onEvent: @escaping (StreamEvent) -> Void
    ) async {
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")

        let messages: [[String: Any]] = conversation.map { message in
            [
                "role": message.role == .user ? "user" : "assistant",
                "content": message.text
            ]
        }
        let payload: [String: Any] = [
            "model": model,
            "max_tokens": 1024,
            "system": systemPrompt,
            "stream": true,
            "messages": messages
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload)

        do {
            let (bytes, response) = try await URLSession.shared.bytes(for: request)
            if let http = response as? HTTPURLResponse, http.statusCode != 200 {
                onEvent(.error(NSError(
                    domain: "ClickySDK.ClaudeAPI",
                    code: http.statusCode,
                    userInfo: [NSLocalizedDescriptionKey: "HTTP \(http.statusCode)"]
                )))
                return
            }
            for try await line in bytes.lines {
                guard line.hasPrefix("data: ") else { continue }
                let dataString = String(line.dropFirst("data: ".count))
                guard let data = dataString.data(using: .utf8),
                      let event = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let type = event["type"] as? String else {
                    continue
                }
                switch type {
                case "content_block_delta":
                    if let delta = event["delta"] as? [String: Any],
                       let text = delta["text"] as? String {
                        onEvent(.textDelta(text))
                    }
                case "message_stop":
                    onEvent(.done)
                    return
                default:
                    break
                }
            }
            onEvent(.done)
        } catch {
            onEvent(.error(error))
        }
    }
}

File: clicky-demo/ClickySDK-Android/clickysdk/src/main/java/com/clicky/sdk/internal/ClaudeApi.kt
package com.clicky.sdk.internal

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

// Minimal Claude streaming client. SSE-based. Mirrors the iOS ClaudeAPI.
//
// PRODUCTION NOTE: this calls api.anthropic.com directly with the key
// passed in via ClickyConfig. For real apps, point it at a proxy URL
// you control so the key isn't embedded in the binary.

internal data class ClaudeMessage(val role: Role, val text: String) {
    enum class Role { USER, ASSISTANT }
}

internal class ClaudeApi(
    private val apiKey: String,
    private val model: String,
) {
    sealed class StreamEvent {
        data class TextDelta(val text: String) : StreamEvent()
        data object Done : StreamEvent()
        data class Error(val message: String) : StreamEvent()
    }

    suspend fun stream(
        systemPrompt: String,
        conversation: List<ClaudeMessage>,
        onEvent: (StreamEvent) -> Unit,
    ) = withContext(Dispatchers.IO) {
        val url = URL("https://api.anthropic.com/v1/messages")
        val conn = url.openConnection() as HttpURLConnection
        try {
            conn.requestMethod = "POST"
            conn.doOutput = true
            conn.setRequestProperty("Content-Type", "application/json")
            conn.setRequestProperty("x-api-key", apiKey)
            conn.setRequestProperty("anthropic-version", "2023-06-01")
            conn.setRequestProperty("Accept", "text/event-stream")

            val messagesArray = JSONArray()
            for (message in conversation) {
                val obj = JSONObject()
                obj.put("role", if (message.role == ClaudeMessage.Role.USER) "user" else "assistant")
                obj.put("content", message.text)
                messagesArray.put(obj)
            }
            val payload = JSONObject()
                .put("model", model)
                .put("max_tokens", 1024)
                .put("system", systemPrompt)
                .put("stream", true)
                .put("messages", messagesArray)
            conn.outputStream.use { it.write(payload.toString().toByteArray()) }

            if (conn.responseCode != 200) {
                onEvent(StreamEvent.Error("HTTP ${conn.responseCode}"))
                return@withContext
            }

            BufferedReader(InputStreamReader(conn.inputStream)).use { reader ->
                while (true) {
                    val line = reader.readLine() ?: break
                    if (!line.startsWith("data: ")) continue
                    val dataString = line.substring("data: ".length)
                    val event = try { JSONObject(dataString) } catch (_: Exception) { continue }
                    when (event.optString("type")) {
                        "content_block_delta" -> {
                            val delta = event.optJSONObject("delta") ?: continue
                            val text = delta.optString("text")
                            if (text.isNotEmpty()) onEvent(StreamEvent.TextDelta(text))
                        }
                        "message_stop" -> {
                            onEvent(StreamEvent.Done)
                            return@withContext
                        }
                    }
                }
                onEvent(StreamEvent.Done)
            }
        } catch (e: Exception) {
            onEvent(StreamEvent.Error(e.message ?: "unknown error"))
        } finally {
            conn.disconnect()
        }
    }
}
